test/boost/flat_mutation_reader_test.cc

```diff
@@ -838,3 +839,65 @@ SEASTAR_THREAD_TEST_CASE(test_reverse_reader_memory_limit) {
     test_with_partition(true);
     test_with_partition(false);
 }
+
+SEASTAR_THREAD_TEST_CASE(test_reverse_reader_reads_in_native_reverse_order) {
+    using namespace tests::data_model;
+
+    tests::reader_concurrency_semaphore_wrapper semaphore;
+    auto permit = semaphore.make_permit();
+
+    auto forward_schema = schema_builder("ks", "cf")
+            .with_column("pk", int32_type, column_kind::partition_key)
+            .with_column("ck", int32_type, column_kind::clustering_key)
```

Comment:
Maybe it would be better to use a composite clustering key just to check the more complex case.

Reply:
Yup, also with reversed types on some of the columns.

Reply:
I think the code path that's responsible for slicing is not affected by thing change @nyh. We will need such test after we implement reversing_readers for memtable and sstables.

Reply:
For now, the change treats the underlying readers as blackboxes so we shouldn't even care about the promoted index as it's the implementation detail of a reader that we're wrapping with reversing_reader.

Reply:
I will rewrite this to use `random_schema`.

Reply:
Why not just reuse all existing test cases for forward readers?

```
diff --git a/test/lib/mutation_source_test.cc b/test/lib/mutation_source_test.cc
index d184cdefd0..54d2770770 100644
--- a/test/lib/mutation_source_test.cc
+++ b/test/lib/mutation_source_test.cc
@@ -1756,6 +1756,22 @@ void run_mutation_source_tests(populate_fn_ex populate) {
                     ms.make_reader(s, std::move(permit), pr, slice, pc, std::move(tr), fwd, mr_fwd));
         });
     });
+
+    run_mutation_reader_tests([populate] (schema_ptr s, const std::vector<mutation>& m, gc_clock::time_point t) -> mutation_source {
+        auto m_rev = reversed(m);
+        auto s_rev = s->make_reversed();
+        return mutation_source([ms = populate(s_rev, m_rev, t)] (schema_ptr s,
+                                              reader_permit permit,
+                                              const dht::partition_range& pr,
+                                              const query::partition_slice& slice,
+                                              const io_priority_class& pc,
+                                              tracing::trace_state_ptr tr,
+                                              streamed_mutation::forwarding fwd,
+                                              mutation_reader::forwarding mr_fwd) {
+            auto rev_slice = slice.make_reversed(); // FIXME: use chained_delegating_reader to keep the new slice alive
+            return ms.make_reader(s, std::move(permit), pr, rev_slice, pc, std::move(tr), fwd, mr_fwd));
+        });
+    });
 }
 
```

Reply:
I'll try this too, but I think there is room for a custom test with random schema too, to make sure we cover composite keys too.